### PR TITLE
PLAT-114355: Adjust Dropdown direction to display properly at the bottom

### PR DESCRIPTION
**Pull request by jeonghee27**
Enact-DCO-1.0-Signed-off-by: Jeonghee Ahn <jeonghee27.ahn@lge.com>

### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [ ] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
[PDM App: Not able to select Dropdown options for the displays at the bottom.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
 - Adjust Dropdown direction by checking the overflow status.
 - If there is a parent node with `overflow` property like `Scroller`, the space is calculated by this node.
 - Add qa-sampler for checking the behavior 

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)
PLAT-114355

### CODE REVIEW

File: samples/sampler/stories/default/Dropdown.js [hunk 1]
```diff
@@ -81,4 +82,56 @@ storiesOf('Agate QA.Dropdown', module)
 		{
 			text: 'The basic Dropdown'
 		}
+	).add(
+		'in Scroller with wrong direction',
```
Comment (MikyungKim) [inline samples/sampler/stories/default/Dropdown.js:86]: repeated in Scroller or with many in Scroller or groups in Scroller?

File: samples/sampler/stories/default/Dropdown.js [hunk 2]
```diff
@@ -81,4 +83,30 @@ storiesOf('Agate QA.Dropdown', module)
 		{
 			text: 'The basic Dropdown'
 		}
+	).add(
+		'groups in Scroller',
```
Comment (MikyungKim) [inline samples/sampler/stories/default/Dropdown.js:87]: maybe group?!

File: Dropdown/Dropdown.js [hunk 3]
```diff
@@ -190,27 +223,16 @@ const DropdownBase = kind({
...
-					return 'up';
-			}
-		},
 		hasChildren: ({children}) => {
 			return children.length > 0;
 		}
 	},
 
-	render: ({buttonClassName, children, css, dropdownButtonClassname, dropdownListClassname, disabled, hasChildren, onClose, onOpen, onSelect, open, selected, skin, transitionContainerClassname, transitionDirection, title, ...rest}) => {
+	render: ({adjustedDirection, buttonClassName, children, css, dropdownListClassname, disabled, hasChildren, onClose, onOpen, onSelect, open, selected, skin, styler, title, ...rest}) => {
```
Comment (MikyungKim) [inline Dropdown/Dropdown.js:231]: Please use `classnames`.

File: Dropdown/Dropdown.js [hunk 4]
```diff
@@ -179,8 +181,39 @@ const DropdownBase = kind({
 
 	computed: {
 		buttonClassName: ({open, styler}) => styler.append({open}),
-		transitionContainerClassname: ({css, open, direction, styler}) => styler.join(css.transitionContainer, {openTransitionContainer: open, upTransitionContainer: direction === 'up'}),
-		dropdownButtonClassname: ({css, direction, styler}) => styler.join(css.dropdownButton, {upDropdownButton: direction === 'up'}),
+		adjustedDirection: ({direction, 'data-spotlight-id': containerId}) => {
+			const calcOverflow = (container, client) => {
+				const KEEPOUT = ri.scale(24); // keep out distance on the edge of the screen
+				const overflow = {
+					isOverTop: client.top - container.height  - KEEPOUT < 0,
```
Comment (MikyungKim) [inline Dropdown/Dropdown.js:188]: two spaces

File: Dropdown/Dropdown.js [hunk 5]
```diff
@@ -179,8 +191,39 @@ const DropdownBase = kind({
...
-		transitionContainerClassname: ({css, open, direction, styler}) => styler.join(css.transitionContainer, {openTransitionContainer: open, upTransitionContainer: direction === 'up'}),
-		dropdownButtonClassname: ({css, direction, styler}) => styler.join(css.dropdownButton, {upDropdownButton: direction === 'up'}),
+		adjustedDirection: ({direction, 'data-spotlight-id': containerId}) => {
+			const calcOverflow = (container, client) => {
+				const KEEPOUT = ri.scale(24); // keep out distance on the edge of the screen
+				const overflow = {
+					isOverTop: client.top - container.height - KEEPOUT < 0,
+					isOverBottom: client.bottom + container.height + KEEPOUT > window.innerHeight
+				};
+				return overflow;
```
Comment (MikyungKim) [inline Dropdown/Dropdown.js:201]: Can we have one empty line above?

File: Dropdown/Dropdown.js [hunk 6]
```diff
@@ -179,8 +191,39 @@ const DropdownBase = kind({
...
+			};
+
+			const adjustDirection = (overflow) => {
+				let adjustedDirection = direction;
+				if (overflow.isOverTop && !overflow.isOverBottom && direction === 'up') {
+					adjustedDirection = 'down';
+				} else if (overflow.isOverBottom && !overflow.isOverTop) {
+					adjustedDirection = 'up';
+				}
+				return adjustedDirection;
```
Comment (MikyungKim) [inline Dropdown/Dropdown.js:211]: Can we have one empty line above?

File: Dropdown/Dropdown.js [hunk 7]
```diff
@@ -179,8 +191,39 @@ const DropdownBase = kind({
...
+					isOverBottom: client.bottom + container.height + KEEPOUT > window.innerHeight
+				};
+				return overflow;
+			};
+
+			const adjustDirection = (overflow) => {
+				let adjustedDirection = direction;
+				if (overflow.isOverTop && !overflow.isOverBottom && direction === 'up') {
+					adjustedDirection = 'down';
+				} else if (overflow.isOverBottom && !overflow.isOverTop) {
```
Comment (MikyungKim) [inline Dropdown/Dropdown.js:208]: Should we add `&& direction === 'down'` ?

File: Dropdown/Dropdown.js [hunk 8]
```diff
@@ -179,9 +191,42 @@ const DropdownBase = kind({
...
+
+			if (containerNode && clientNode) {
+				const containerNodeRect = containerNode.getBoundingClientRect();
+				const clientNodeRect = clientNode.getBoundingClientRect();
+				return adjustDirection(calcOverflow(containerNodeRect, clientNodeRect));
+			}
+
+			return direction;
+		},
+
```
Comment (MikyungKim) [inline Dropdown/Dropdown.js:228]: Could you remove this line?

**Review by MikyungKim — APPROVED** [review]
LGTM

**Comment by 0x64** [comment]
@jeonghee27 Could you please resolve conflicts?

**Pull request merged**